src/schemas/ProgramDonorSummary/resolvers/index.ts

```diff
@@ -3,13 +3,76 @@ import { GlobalGqlContext } from 'app';
 import { createEsClient } from 'services/elasticsearch';
 import programDonorSummaryEntriesResolver from './summaryEntries';
 import programDonorSummaryStatsResolver from './summaryStats';
+import { GraphQLFieldResolver } from 'graphql';
+import egoTokenUtils from 'utils/egoTokenUtils';
+import { AuthenticationError, ApolloError } from 'apollo-server-express';
+
+class UnauthorizedError extends ApolloError {
+  extensions = {
+    code: 'UNAUTHORIZED',
+  };
+}
+
+const resolveWithProgramAuth = <ResolverType = GraphQLFieldResolver<unknown, unknown, unknown>>(
+  context: GlobalGqlContext,
+  args: { programShortName: string },
+  resolver: ResolverType,
+): ResolverType => {
+  const { egoToken } = context;
+  const {
+    decodeToken,
+    isExpiredToken,
+    getPermissionsFromToken,
+    isValidJwt,
+    canReadProgramData,
+    canReadProgram,
+  } = egoTokenUtils;
+
+  if (egoToken) {
+    let decodedToken: ReturnType<typeof decodeToken>;
+    try {
+      decodedToken = decodeToken(egoToken);
+    } catch (err) {
+      throw new AuthenticationError(err);
+    }
+
+    const isExpired = isExpiredToken(decodedToken);
+    const permissions = getPermissionsFromToken(egoToken);
+    const hasPermission =
+      canReadProgram({
+        permissions,
+        programId: args.programShortName,
+      }) ||
+      canReadProgramData({
+        permissions,
+        programId: args.programShortName,
+      });
+
+    const authorized =
+      egoToken && isValidJwt(egoToken) && !isExpiredToken(decodedToken) && hasPermission;
```

Comment:
could use the const and check `!isExpired`...??

Reply:
yes! :P me derped